### PR TITLE
donate prototype from krancour/brigade-github-gateway-prototype

### PR DESCRIPTION
**Pull request by krancour**
This PR donates to Brigade 2 compatible GitHub gateway prototype from krancour/brigade-github-gateway-prototype to an official brigadecore repo.

I probably could have just pushed this as an "Initial commit," but felt better having it scrutinized.

@vdice want to walk through it tomorrow?

To anyone familiar with Brigade 2 and the old GitHub gateway, this is all very simple and recognizable code.

### CODE REVIEW

**Comment by vdice** [comment]
> @vdice want to walk through it tomorrow?

Definitely!

File: receiver/internal/webhooks/service.go [hunk 1]
```diff
@@ -0,0 +1,872 @@
...
+		} else if refSubmatches :=
+			tagRefRegex.FindStringSubmatch(*pe.Ref); len(refSubmatches) == 2 {
+			shortTitle = fmt.Sprintf("tag: %s", refSubmatches[1])
+			longTitle = shortTitle
+		}
+	}
+	return shortTitle, longTitle
+}
+
+// getTitlesFromPushEvent extracts human-readable event titles from a
```
Comment (vdice) [inline receiver/internal/webhooks/service.go:751]:
```suggestion
// getTitlesFromPR extracts human-readable event titles from a
```

File: receiver/internal/webhooks/service_test.go [hunk 2]
```diff
@@ -0,0 +1,1696 @@
...
+						Commit: testSHA,
+						Ref:    testBranch,
+					},
+					*event.Git,
+				)
+			},
+		},
+
+		{
+			name:      "check_run event; action is rerequested",
```
Comment (vdice) [inline receiver/internal/webhooks/service_test.go:159]: I'm curious as to the difference in tested functionality between this and the prior, generic action test?  (Similarly for the check_suite tests below).  Should there be other/extra validations for when the action is `rerequested`?  Or, I may have missed them :)

**Review by vdice — COMMENTED** [review]
Awesome work!  I like the split into monitor and receiver components; makes sense.  Also, huge kudos for running through all of GH's supported events, adding api links, creating the super helpful table in the README, etc.

Most comments are minor and others would be totally fine as follow-ups (if even worthy).  I think we just need to add the chart linting target and resolve the `/brig run` vs `/brig check` issue comment question before merging...

File: Makefile [hunk 3]
```diff
@@ -0,0 +1,190 @@
...
+	VERSION            := $(GIT_VERSION)
+	MUTABLE_DOCKER_TAG := edge
+endif
+
+IMMUTABLE_DOCKER_TAG := $(VERSION)
+
+################################################################################
+# Tests                                                                        #
+################################################################################
+
```
[the remark this answers is not in the document]
Comment (krancour) [reply Makefile:89]: In progress. I was about to push it when we spoke earlier and then I forgot!

File: receiver/internal/webhooks/service.go [hunk 4]
```diff
@@ -0,0 +1,872 @@
...
+	case *github.IssueCommentEvent:
+		brigadeEvent.Type = fmt.Sprintf("issue_comment:%s", e.GetAction())
+		brigadeEvent.Qualifiers = map[string]string{
+			"repo": e.GetRepo().GetFullName(),
+		}
+		// Under a very specific set of conditions, we will request a check suite to
+		// run in response to this comment.
+		//
+		// 1. The issue in question is a PR
+		// 2. The comment contains "/brig check" (case insensitive)
```
[the remark this answers is not in the document]
Comment (krancour) [reply receiver/internal/webhooks/service.go:358]: Yes. Nice catch. Will fix.

File: Makefile [hunk 5]
```diff
@@ -0,0 +1,198 @@
...
+################################################################################
+# Build / Publish                                                              #
+################################################################################
+
+.PHONY: build
+build: build-images
+
+.PHONY: build-images
+build-images: build-receiver
```
Comment (vdice) [inline Makefile:120]: Oh, we need to add `build-monitor`, right?

Comment (krancour) [reply Makefile]: Fixed

File: receiver/internal/webhooks/service_test.go (hunk 2 above)
Comment (krancour) [reply receiver/internal/webhooks/service_test.go:159]: Oh... this was a good catch, @vdice.

When I'm constructing these table-driven tests, I usually have the tests open in one editor window and the source _under_ test open in another. I frequently re-run the test to see if I've covered most nooks and crannies of the source under test-- i.e. different conditionals. So I did that here-- perhaps blindly-- failing to note that the conditional in question doesn't do anything different than what would happen if it weren't there...

But this is a mistake in the code under test and not in the test.

```
brigadeEvent.SourceState = &core.SourceState{
	State: map[string]string{
		"tracking":       "true",
		"installationID": strconv.FormatInt(e.GetInstallation().GetID(), 10),
		"owner":          e.GetRepo().GetOwner().GetLogin(),
		"repo":           e.GetRepo().GetName(),
		"headSHA":        e.GetCheckRun().GetCheckSuite().GetHeadSHA(),
	},
}
if e.GetAction() == "rerequested" {
	if s.shouldEmit(brigadeEvent.Type) {
		if events, err = s.eventsClient.Create(ctx, brigadeEvent); err != nil {
			return events, errors.Wrap(err, "error emitting event(s) into Brigade")
		}
	}
	return events, nil // We're done
}
```

Two mistakes actually...

First, this is an artifact of the pre-monitor prototype where I just created the event and then immediately launched a goroutine to monitor it right here in the same process. (That wasn't a good approach because it's not resilient to this process failing. And that's why I ended up creating source state and the monitor component.)

The second mistake is that the conditional should be up higher, so the thing that is conditional is whether we're including the source state that prompts the monitor to follow up on the event.

Bottom line is I am fixing this.

Thanks!

Comment (krancour) [reply receiver/internal/webhooks/service_test.go:159]: This is also applicable to check_suite events.

Comment (krancour) [reply receiver/internal/webhooks/service_test.go:159]: And now I suppose there's no reason to have separate test cases either (since there's no `else`). I can just always test the case that gets me into that conditional.

Comment (krancour) [reply receiver/internal/webhooks/service_test.go]: All fixed up in ee09f623a045a545d5b3b6a15ff1cca39ed75676

Comment (vdice) [reply receiver/internal/webhooks/service_test.go]: Nice, glad this prompted the code fix!  Thanks for the explanation/background as well.

File: receiver/internal/webhooks/service.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (krancour) [reply receiver/internal/webhooks/service.go]: Fixed in f9037fa91a53bcc74b6bfe52993abca43755b80e

File: charts/brigade-github-gateway/values.yaml [hunk 6]
```diff
@@ -0,0 +1,230 @@
+## Default values for brigade-github-gateway
+## This is a YAML-formatted file.
+## Declare variables to be passed into your templates.
+
+brigade:
+  ## Address of your Brigade 2 API server, including leading protocol (http://
+  ## or https://)
+  apiAddress: https://brigade-apiserver.brigade.svc.cluster.local
+  ## API token belonging to a Brigade 2 service account
+  apiToken: 0da4dab0b1f64c28aaebd652601ff5a3uBbDGVkqDPz8c9evn86AHvIwRUk9b5TAMllKrbLxsLwmcGgXwD87VEpcMm44rXHX1o6IZKk29GNxBsESAYHQKoOAy8qY28D85vGsMx5MeosVNFthFpYzIeOugiuXsK7vM9lD4xIdTfBRQMGTQjVQ1BcDF4vUFluV5OrBaOwFlLNiCoP9U4bq7EPIWRYppIG7t2D2QKzZ7n6S73PnCH7dE8xp2aaeBmaV
```
[the remark this answers is not in the document]
Comment (krancour) [reply charts/brigade-github-gateway/values.yaml:10]: NOPE. lol. Fixing.

File: charts/brigade-github-gateway/values.yaml [hunk 7]
```diff
@@ -0,0 +1,230 @@
...
+  ## Whether to ignore cert warning from the API server
+  apiIgnoreCertWarnings: true
+
+## You must create a GitHub App (https://github.com/settings/apps/new)
+## corresponding to this Brigade GitHub Gateway.
+github:
+  ## This is the unique ID of your GitHub App. It is assigned by GitHub and can
+  ## be retrieved from your GitHub App's main page after you have created it.
+  appID: 99005
```
[the remark this answers is not in the document]
Comment (krancour) [reply charts/brigade-github-gateway/values.yaml:19]: NOPE. lol. Fixing.

File: charts/brigade-github-gateway/values.yaml [hunk 8]
```diff
@@ -0,0 +1,230 @@
...
+  ## A PEM-formatted x509 key for your GitHub App. This value should begin with
+  ## "-----BEGIN RSA PRIVATE KEY-----" and end with
+  ## "-----END RSA PRIVATE KEY-----"
+  apiKey: |
```
[the remark this answers is not in the document]
Comment (krancour) [reply charts/brigade-github-gateway/values.yaml:23]: Again... NOPE.

Also, note that none of these credentials are currently valid, so there's no exposure.

[the remark this answers is not in the document]
Comment (krancour) [reply charts/brigade-github-gateway/values.yaml:23]: Fixed in 4992c4e3aad5d0f6e49d7e5919e1a042fa7b4e2d

File: charts/brigade-github-gateway/values.yaml (hunk 7 above)
[the remark this answers is not in the document]
Comment (krancour) [reply charts/brigade-github-gateway/values.yaml]: Fixed in 4992c4e3aad5d0f6e49d7e5919e1a042fa7b4e2d

File: charts/brigade-github-gateway/values.yaml (hunk 6 above)
[the remark this answers is not in the document]
Comment (krancour) [reply charts/brigade-github-gateway/values.yaml:10]: Fixed in 4992c4e3aad5d0f6e49d7e5919e1a042fa7b4e2d

File: internal/os/os.go [hunk 9]
```diff
@@ -0,0 +1,130 @@
...
+import (
+	"os"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/pkg/errors"
+)
+
+// GetEnvVar retrieves the value of an environment variable having the specified
```
[the remark this answers is not in the document]
Comment (krancour) [reply internal/os/os.go:12]: This is true. We also have some other code in common-- packages `files`, `signals`, and `version`.

My mind could be changed, but my initial thought was that none of these rose to a level of being substantial enough to warrant becoming their own thing.

File: monitor/events.go [hunk 10]
```diff
@@ -0,0 +1,442 @@
...
+		&core.LogsSelector{
+			Job: job.Name,
+		},
+		nil,
+	)
+	if err != nil {
+		return "", err
+	}
+	// Arbitrarily limiting to 1000 log lines because we don't want to blow
+	// out the heap if a Job has produced an enormous amount of logs
```
[the remark this answers is not in the document]
Comment (krancour) [reply monitor/events.go:417]: That is a great idea. Adding...

[the remark this answers is not in the document]
Comment (krancour) [reply monitor/events.go:417]: Fixed in b10f42ecc926a9cbf31b6f5c3c8321b7e35a8d38

File: receiver/internal/http/server.go [hunk 11]
```diff
@@ -0,0 +1,135 @@
...
+type ServerConfig struct {
+	// Port specifies the port the server should bind to / listen on.
+	Port int
+	// TLSEnabled specifies whether the server should serve HTTP (false) or HTTPS
+	// (true).
+	TLSEnabled bool
+	// TLSCertPath is the path to a PEM-encoded x509 certificate that can be used
+	// for serving HTTPS.
+	TLSCertPath string
+	// TLSCertPath is the path to a PEM-encoded x509 private key that can be used
```
[the remark this answers is not in the document]
Comment (krancour) [reply receiver/internal/http/server.go]: Fixed

File: receiver/internal/webhooks/handler.go [hunk 12]
```diff
@@ -0,0 +1,66 @@
+package webhooks
+
+import (
+	"encoding/json"
+	"log"
+	"net/http"
+
+	// nolint: lll
```
[the remark this answers is not in the document]
Comment (krancour) [reply receiver/internal/webhooks/handler.go:8]: Nope... leftover from before some things got split out of this file and changed what the imports were.

Fixing.

[the remark this answers is not in the document]
Comment (krancour) [reply receiver/internal/webhooks/handler.go]: Fixed.

File: monitor/events.go [hunk 13]
```diff
@@ -0,0 +1,442 @@
...
+		case <-ticker.C:
+		case <-ctx.Done():
+			return nil
+		}
+	}
+}
+
+func (m *monitor) createCheckRun(
+	ctx context.Context,
+	installationID int64,
```
[the remark this answers is not in the document]
Comment (krancour) [reply monitor/events.go:253]: I think I'd rather not here if it's ok with you. My rationale is that I like doing that when there are a lot of optional bits of information. It works nicely for things like `config`, `opts`, or `selector`. It's also a nice trick anywhere where I anticipate the list of arguments growing over time, but I want to avoid changing public function signatures. But with a private function where all where all the args are also required, bundling them all into a struct provides less benefit and (if we're being super careful) also might prompt us to write extra code to guard against missing values.

File: charts/brigade-github-gateway/values.yaml [hunk 14]
```diff
@@ -0,0 +1,230 @@
...
+    -----END RSA PRIVATE KEY-----
+
+## All settings for the receiver
+receiver:
+
+  replicas: 1
+
+  ## Host should be set accurately for a variety of reasons. This is used in
+  ## ingress resources and cert generation.
+  host: localhost:31700
```
[the remark this answers is not in the document]
Comment (krancour) [reply charts/brigade-github-gateway/values.yaml:33]: It should be the gateway's DNS / host name. I'll add more docs about this.

Brigade itself might need the same treatment.

File: monitor/events.go (hunk 13 above)
[the remark this answers is not in the document]
Comment (vdice) [reply monitor/events.go:253]: 👍 sgtm

File: charts/brigade-github-gateway/values.yaml (hunk 6 above)
[the remark this answers is not in the document]
Comment (vdice) [reply charts/brigade-github-gateway/values.yaml:10]: @krancour can you double-check the `apiToken` line?  I think a value still remains here.

File: README.md [hunk 15]
```diff
@@ -0,0 +1,415 @@
...
+    --description brigade-github-gateway
+```
+
+Make note of the __token__ returned. This value will be used in another step.
+_It is your only opportunity to access this value, as Brigade does not save it._
+
+Authorize this service account to read all events and to create new ones:
+
+```console
+$ bin/brig-darwin-amd64 role grant READER \
```
Comment (vdice) [inline README.md:121]: Shall we update this and the next line to use `brig` as opposed to `bin/brig-darwin-amd64`?

File: charts/brigade-github-gateway/values.yaml (hunk 14 above)
[the remark this answers is not in the document]
Comment (krancour) [reply charts/brigade-github-gateway/values.yaml]: Fixed in eb5a51469cb6bcfe5406e60595f454f0352cacfe

File: README.md (hunk 15 above)
Comment (krancour) [reply README.md:121]: Good idea... since `bin/brig-darwin-amd64` is very much just an artifact of the fact that we're simultaneously hacking on Brigade v2 itself.

Will fix.

Comment (krancour) [reply README.md]: Fixed in a9c5068c9f74a0cc75d03751abb0e3f29bcda08b

**Review by vdice — APPROVED** [review]
LGTM! 👏

**Comment by krancour** [comment]
@vdice as we discussed offline, I'll hold off on the merge until I've updated infra with secrets needed to push edge images.
